### PR TITLE
[MRG+1] Added axis limit check for non-finite values

### PR DESCRIPTION
**Pull request by bcongdon**
Addresses #7460.

Raises a `ValueError` if non-finite values like `np.nan` or `np.inf` are passed as arguments to `set_xlim` or `set_ylim`.

### CODE REVIEW

**Comment by dstansby** [comment]
Should this raise a warning instead of an error? Currently invalid limits on a log axis raises a warning (#7367), which I think is best. Either way either a warning or error should be raised consistently.

**Comment by bcongdon** [comment]
Yeah, that sounds like the better approach. Passing an invalid value to `plt.xlim` results in undefined behavior, but a plot can still be drawn.

**Comment by tacaswell** [comment]
The consensus in #7460 seemed to be in favor of errors, not warnings.  It seems that warnings either just annoy people (because they are doing the thing for a good reason) or are ignored.

I think this will still re-set the limits, which was the original complaint.

**Comment by anntzer** [comment]
Please note https://github.com/matplotlib/matplotlib/issues/7733#issuecomment-272325978.

**Comment by bcongdon** [comment]
I've looked over the discussion on #7733 and related issues. Still a bit confused on how to proceed. I think the issues are somewhat unrelated in that the solution for #7735 - letting invalid values pass through - is actually the cause of the issue this PR addresses (#7460).

I've reset back to the commit that raises errors instead of warnings.

**Comment by dstansby** [comment]
I think the consensus is to raise an error, which sounds good to me (see #7460). @bcongdon could you rebase this on to the current master branch? Otherwise it looks good to go.

**Comment by bcongdon** [comment]
Rebased. 👍

File: lib/matplotlib/tests/test_axes.py [hunk 1]
```diff
@@ -4931,3 +4931,15 @@ def test_bar_single_height():
     ax.bar(range(4), 1)
     # Check that a horizontal chart with one width works
     ax.bar(0, 1, bottom=range(4), width=1, orientation='horizontal')
+
+
+def test_invalid_axis_limits():
```
Comment (dstansby) [inline lib/matplotlib/tests/test_axes.py]: One last thing, could a `@cleanup` decorator go above this line. Otherwise looks good!

Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Not needed on current `master`.

File: lib/matplotlib/axes/_base.py [hunk 2]
```diff
@@ -2875,6 +2875,11 @@ def set_xlim(self, left=None, right=None, emit=True, auto=False, **kw):
         if right is not None:
             right = self.convert_xunits(right)
 
+        if ((left is not None and not np.isfinite(left)) or
+                (right is not None and not np.isfinite(right))):
+            raise ValueError("xlim limits must be finite. "
+                             "instead, found: (%s, %s)" % (left, right))
```
Comment (QuLogic) [inline lib/matplotlib/axes/_base.py]: 'xlim limits' is a bit redundant. Also, 'instead' comes after a period so should be capitalized. Something like 'Specified x limits must be finite; instead found: (%s, %s)'.

File: lib/matplotlib/axes/_base.py [hunk 3]
```diff
@@ -3169,6 +3174,11 @@ def set_ylim(self, bottom=None, top=None, emit=True, auto=False, **kw):
         if top is not None:
             top = self.convert_yunits(top)
 
+        if ((top is not None and not np.isfinite(top)) or
+                (bottom is not None and not np.isfinite(bottom))):
+            raise ValueError("ylim limits must be finite. "
+                             "instead, found: (%s, %s)" % (top, bottom))
```
Comment (QuLogic) [inline lib/matplotlib/axes/_base.py]: Same comment here, but also, 'bottom' should come before 'top'.

File: lib/matplotlib/axes/_base.py (hunk 2 above)
Comment (bcongdon) [inline lib/matplotlib/axes/_base.py]: Alright, sounds good. That redundancy bothered me too, but I couldn't think of the right verbiage to describe the warning. Thanks!

File: lib/matplotlib/axes/_base.py [hunk 4]
```diff
@@ -3176,7 +3176,7 @@ def set_ylim(self, bottom=None, top=None, emit=True, auto=False, **kw):
 
         if ((top is not None and not np.isfinite(top)) or
                 (bottom is not None and not np.isfinite(bottom))):
-            raise ValueError("ylim limits must be finite. "
+            raise ValueError("Specified y limits must be finite; "
                              "instead, found: (%s, %s)" % (top, bottom))
```
Comment (QuLogic) [inline lib/matplotlib/axes/_base.py]: Still need to swap `top` and `bottom`.

Comment (bcongdon) [inline lib/matplotlib/axes/_base.py]: Whoops, my bad.

**Comment by QuLogic** [comment]
@tacaswell To what documentation are you referring in your review? Can it be dismissed?

**Comment by QuLogic** [comment]
Ping @tacaswell?

**Pull request merged**